diff -ruabB ./old/freemarker/ext/beans/BeansWrapper.java ./new/freemarker/ext/beans/BeansWrapper.java
--- ./old/freemarker/ext/beans/BeansWrapper.java	2008-05-05 19:56:36.000000000 +0300
+++ ./new/freemarker/ext/beans/BeansWrapper.java	2010-06-07 17:11:25.744592485 +0300
@@ -1146,7 +1146,7 @@
                     }
                 }
             }
-            if(exposureLevel < EXPOSE_PROPERTIES_ONLY)
+            if(exposureLevel < EXPOSE_PROPERTIES_ONLY && mda != null)
             {
                 for(int i = mda.length - 1; i >= 0; --i)
                 {
